lib/srv/ctx_test.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"testing"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCheckFileCopyingAllowed(t *testing.T) {
+	srv := newMockServer(t)
+	ctx := newTestServerContext(t, srv, nil)
+
+	tests := []struct {
+		name                 string
+		nodeAllowFileCopying bool
+		roles                []types.Role
+		expectedErr          error
+	}{
+		{
+			name:                 "node disallowed",
+			nodeAllowFileCopying: false,
+			roles: []types.Role{
+				&types.RoleV5{
+					Kind: types.KindNode,
+				},
+			},
+			expectedErr: errNodeFileCopyingNotPermitted,
+		},
+		{
+			name:                 "node allowed",
+			nodeAllowFileCopying: true,
+			roles: []types.Role{
+				&types.RoleV5{
+					Kind: types.KindNode,
+				},
+			},
+			expectedErr: nil,
+		},
+		{
+			name:                 "role disallowed",
+			nodeAllowFileCopying: true,
+			roles: []types.Role{
+				&types.RoleV5{
+					Kind: types.KindNode,
+					Spec: types.RoleSpecV5{
+						Options: types.RoleOptions{
+							FileCopying: types.NewBoolOption(false),
+						},
+					},
+				},
+			},
+			expectedErr: errRoleFileCopyingNotPermitted,
+		},
+		{
+			name:                 "role allowed",
+			nodeAllowFileCopying: true,
+			roles: []types.Role{
+				&types.RoleV5{
+					Kind: types.KindNode,
+					Spec: types.RoleSpecV5{
+						Options: types.RoleOptions{
+							FileCopying: types.NewBoolOption(true),
+						},
+					},
+				},
+			},
+			expectedErr: nil,
+		},
+		{
+			name:                 "conflicting roles",
+			nodeAllowFileCopying: true,
+			roles: []types.Role{
+				&types.RoleV5{
+					Kind: types.KindNode,
+					Spec: types.RoleSpecV5{
+						Options: types.RoleOptions{
+							FileCopying: types.NewBoolOption(true),
+						},
+					},
+				},
+				&types.RoleV5{
+					Kind: types.KindNode,
+					Spec: types.RoleSpecV5{
+						Options: types.RoleOptions{
+							FileCopying: types.NewBoolOption(false),
+						},
+					},
+				},
+			},
+			expectedErr: errRoleFileCopyingNotPermitted,
+		},
+	}
+
+	for _, tt := range tests {
+		ctx.allowFileCopying = tt.nodeAllowFileCopying
```

Comment:
I would put all of this in a subtest with `t.Run` so it's clear which test fails if one of these starts failing.

Reply:
Done